src/Shell/Task/ModelTask.php

```diff
@@ -395,17 +441,74 @@ public function findBelongsToMany($model, array $associations)
                 $assocTable = substr($otherTableName, 0, $otherOffset);
             }
             if ($assocTable && in_array($assocTable, $tables)) {
-                $habtmName = $this->_camelize($assocTable);
-                $assoc = [
-                    'alias' => $habtmName,
-                    'foreignKey' => $foreignKey,
-                    'targetForeignKey' => $this->_modelKey($habtmName),
-                    'joinTable' => $otherTableName
-                ];
-                if ($assoc && $this->plugin) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+                if (!in_array($otherTableName, $tables)) {
+                    continue;
+                }
+                $assocTableObject = TableRegistry::get($otherTableName);
+
+                $thisForeignKey = $primaryKey;
+                foreach ($thisForeignKey as $i => $fieldName) {
+                    if ($fieldName === 'id') {
+                        $thisForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
+                    }
+                }
+                $otherForeignKeys = $assocTableObject->schema()->columns();
+                foreach ($otherForeignKeys as $i => $fieldName) {
+                    if ($fieldName !== 'parent_id' && (
+                        strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName))
+                    ) {
+                        unset($otherForeignKeys[$i]);
+                    }
+                }
+                $thisForeignKey = array_values($thisForeignKey);
+                $otherForeignKeys = array_values($otherForeignKeys);
+                $thisForeignKeyIntersection = array_intersect($thisForeignKey, $otherForeignKeys);
+
+                $reverseForeignKey = false;
+                $reverseForeignKeyIntersection = false;
+                $otherForeignKeysDiff = array_diff($otherForeignKeys, $thisForeignKey);
+                foreach ($otherForeignKeysDiff as $reverseForeignKey) {
```

Comment:
You are right, detection will only work if you follow conventions where there is at least one composite part named id on both sides of HABTM.

#### Example (PK = (Lines.line_number, Lines.file_id) => detection not working)
*Note that the FOREIGN KEY definitions are here to clear up intentions. Bake does not respect those but uses conventions.*
```sql
CREATE TABLE files
(
  id serial NOT NULL,
  name character varying(255) NOT NULL,
  CONSTRAINT files_pkey PRIMARY KEY (id)
);
CREATE TABLE lines
(
  file_id integer NOT NULL,
  line_number integer NOT NULL,
  CONSTRAINT lines_pkey PRIMARY KEY (file_id, line_number),
  CONSTRAINT lines_file_id FOREIGN KEY (file_id)
      REFERENCES files (id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE
);
CREATE TABLE errors
(
  file_id integer NOT NULL,
  id integer NOT NULL,
  CONSTRAINT errors_pkey PRIMARY KEY (file_id, id),
  CONSTRAINT errors_file_id FOREIGN KEY (file_id)
      REFERENCES files (id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE
);
CREATE TABLE errors_lines
(
  file_id integer NOT NULL,
  line_number integer NOT NULL,
  error_id integer NOT NULL,
  CONSTRAINT errors_lines_pkey PRIMARY KEY (file_id, line_number, error_id),
  CONSTRAINT errors_lines_file_id FOREIGN KEY (file_id)
      REFERENCES files (id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE,
  CONSTRAINT errors_lines_file_id_error_id FOREIGN KEY (file_id, error_id)
      REFERENCES errors (file_id, id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE,
  CONSTRAINT errors_lines_file_id_line_number FOREIGN KEY (file_id, line_number)
      REFERENCES lines (file_id, line_number) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE
);
```
**Results into...**
#### FilesTable
```php
        $this->primaryKey('id');

        $this->hasMany('Errors', [
            'foreignKey' => ['file_id']
        ]);
        $this->hasMany('ErrorsLines', [
            'foreignKey' => ['file_id']
        ]);
        $this->hasMany('Lines', [
            'foreignKey' => ['file_id']
        ]);
```
#### ErrorsTable
```php
        $this->primaryKey(['file_id', 'id']);

        $this->belongsTo('Files', [
            'foreignKey' => ['file_id'],
            'joinType' => 'INNER'
        ]);
```
#### LinesTable
```php
        $this->primaryKey(['file_id', 'line_number']);

        $this->belongsTo('Files', [
            'foreignKey' => ['file_id'],
            'joinType' => 'INNER'
        ]);
```
#### ErrorsLinesTable
```php
        $this->primaryKey(['file_id', 'line_number', 'error_id']);

        $this->belongsTo('Files', [
            'foreignKey' => ['file_id'],
            'joinType' => 'INNER'
        ]);
        $this->belongsTo('Errors', [
            'foreignKey' => ['file_id', 'error_id'],
            'joinType' => 'INNER'
        ]);
```

**To get this to work the only change required is in lines and errors_lines table:**

#### Example (PK = (Lines.id, Lines.file_id) => detection not working)

```sql
CREATE TABLE lines
(
  file_id integer NOT NULL,
  id integer NOT NULL,
  CONSTRAINT lines_pkey PRIMARY KEY (file_id, id),
  CONSTRAINT lines_file_id FOREIGN KEY (file_id)
      REFERENCES files (id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE
);
CREATE TABLE errors_lines
(
  file_id integer NOT NULL,
  line_id integer NOT NULL,
  error_id integer NOT NULL,
  CONSTRAINT errors_lines_pkey PRIMARY KEY (file_id, line_id, error_id),
  CONSTRAINT errors_lines_file_id FOREIGN KEY (file_id)
      REFERENCES files (id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE,
  CONSTRAINT errors_lines_file_id_error_id FOREIGN KEY (file_id, error_id)
      REFERENCES errors (file_id, id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE,
  CONSTRAINT errors_lines_file_id_line_id FOREIGN KEY (file_id, line_id)
      REFERENCES lines (file_id, id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE CASCADE
);
```

**Only then bake detects the HABTM association...**

#### ErrorsTable
```php
        $this->primaryKey(['file_id', 'id']);

        $this->belongsTo('Files', [
            'foreignKey' => ['file_id'],
            'joinType' => 'INNER'
        ]);
        $this->belongsToMany('Lines', [
            'foreignKey' => ['file_id', 'error_id'],
            'targetForeignKey' => ['file_id', 'line_id'],
            'joinTable' => 'errors_lines'
        ]);
```
#### LinesTable
```
        $this->primaryKey(['file_id', 'id']);

        $this->belongsTo('Files', [
            'foreignKey' => ['file_id'],
            'joinType' => 'INNER'
        ]);
        $this->belongsToMany('Errors', [
            'foreignKey' => ['file_id', 'line_id'],
            'targetForeignKey' => ['file_id', 'error_id'],
            'joinTable' => 'errors_lines'
        ]);
    }
```
#### ErrorsLinesTable
```php
        $this->primaryKey(['file_id', 'line_id', 'error_id']);

        $this->belongsTo('Files', [
            'foreignKey' => ['file_id'],
            'joinType' => 'INNER'
        ]);
        $this->belongsTo('Lines', [
            'foreignKey' => ['file_id', 'line_id'],
            'joinType' => 'INNER'
        ]);
        $this->belongsTo('Errors', [
            'foreignKey' => ['file_id', 'error_id'],
            'joinType' => 'INNER'
        ]);
```
IMHO: If you want non conventional primary / foreign key binding to work, bake won't do that for you unless we drop the conventions and use real, schema based, foreign keys.